File: java/src/main/java/br/com/vps/consulting/b2b/management/partner/infrastructure/mapper/PartnerMapper.java
package br.com.vps.consulting.b2b.management.partner.infrastructure.mapper;

import br.com.vps.consulting.b2b.management.partner.domain.Partner;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerCreditEntity;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerEntity;

import java.math.BigDecimal;

public final class PartnerMapper {

    private PartnerMapper(){}

    public static Partner toDomain(final PartnerEntity partner, final PartnerCreditEntity credit) {
        final var creditLimit = credit != null ? credit.getCreditLimit() : BigDecimal.ZERO;
        return Partner.builder()
                .id(PartnerId.from(partner.getId()))
                .name(partner.getName())
                .document(partner.getDocument())
                .creditLimit(creditLimit)
                .createdAt(partner.getCreatedAt())
                .build();
    }

    public static PartnerEntity toEntity(final Partner partner) {
        return PartnerEntity.builder()
                .id(partner.getId().value())
                .name(partner.getName())
                .document(partner.getDocument())
                .createdAt(partner.getCreatedAt())
                .build();
    }

}

File: kotlin/src/main/kotlin/br/com/vps/consulting/b2b/management/partner/infrastructure/mapper/PartnerMapper.kt
package br.com.vps.consulting.b2b.management.partner.infrastructure.mapper

import br.com.vps.consulting.b2b.management.partner.domain.Partner
import br.com.vps.consulting.b2b.management.partner.domain.PartnerBalance
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerBalanceEntity
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerEntity
import br.com.vps.consulting.b2b.management.shared.core.vo.Money
import java.math.BigDecimal
import java.time.Instant
import java.util.UUID

fun PartnerEntity.toDomain(): Partner = Partner.with(
    id = PartnerId.from(id),
    name = name,
    document = document,
    createdAt = createdAt,
)

fun Partner.toEntity(): PartnerEntity = PartnerEntity(
    id = id.value,
    name = name,
    document = document,
    createdAt = createdAt,
)

fun PartnerBalanceEntity.toDomain(): PartnerBalance = PartnerBalance(
    partnerId = PartnerId.from(partnerId),
    totalBalance = Money.of(totalBalance),
    availableBalance = Money.of(availableBalance),
    updatedAt = updatedAt,
)

fun zeroBalanceEntityFor(partnerId: UUID): PartnerBalanceEntity = PartnerBalanceEntity(
    partnerId = partnerId,
    totalBalance = BigDecimal.ZERO,
    availableBalance = BigDecimal.ZERO,
    updatedAt = Instant.now(),
)
